### PR TITLE
[ENHANCEMENT] Add z-loss

### PR DESCRIPTION
**Pull request by wdevazelhes**
z-loss regularization was shown to stabilize training by preventing logits in the last layer to explode: https://arxiv.org/abs/2309.14322

This PR: 
- allows one to add zloss regularization
- since the total loss is not cross entropy anymore but cross entropy + zloss, this PR also allows one to additionally log the original cross-entropy and the z-loss

### CODE REVIEW

**Comment by wdevazelhes** [comment]
After more careful consideration, seems it could be better to implement z-loss similarly to the z-loss for MOE, be it for the loss itself, but also for the logging :+1:
